### PR TITLE
osd/PG: fix update_snap_map remove fails if already removed or does not exist

### PR DESCRIPTION
**Pull request by gmayyyha**
Fixes: http://tracker.ceph.com/issues/20974

Signed-off-by: Yanhu Cao <gmayyyha@gmail.com>

### CODE REVIEW

File: src/osd/PG.cc [hunk 1]
```diff
@@ -3374,7 +3374,10 @@ void PG::update_snap_map(
 	int r = snap_mapper.remove_oid(
 	  i->soid,
 	  &_t);
-	assert(r == 0);
+	    if (!(r == 0 || r == -ENOENT)) {
```
Comment (Liuchang0812) [inline src/osd/PG.cc]: Nit, indent

**Comment by liewegas** [comment]
@Liuchang0812 I looked through part of that log and it didn't look like it was ever removed, and then it got ENOENT.  This makes me think there is a problem in reading the entry or maybe it wasn't recorded, or maybe something corrupted it...

File: src/osd/PG.cc (hunk 1 above)
Comment (gmayyyha) [inline src/osd/PG.cc]: @Liuchang0812 done.

**Comment by amitkumar50** [comment]
@gmayyyha  https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#title-of-pull-requests-and-title-of-commits
Title is too long. "Submitting patch Ceph says max of 50 characters"

**Comment by liewegas** [comment]
@Liuchang0812 ping... do we have reason to believe that it's okay to get ENOENT in this case?  I suspect there is an error elsewhere.

File: src/osd/PG.cc [hunk 2]
```diff
@@ -3374,7 +3374,10 @@ void PG::update_snap_map(
 	int r = snap_mapper.remove_oid(
 	  i->soid,
 	  &_t);
-	assert(r == 0);
+	if (!(r == 0 || r == -ENOENT)) {
```
Comment (amitkumar50) [inline src/osd/PG.cc]: @gmayyyha I do believe with @liewegas that there are likely less chances of '  if (got.empty())    return -ENOENT;` is failing to get keys.

Comment (gmayyyha) [reply src/osd/PG.cc]: @amitkumar50, @liewegas I see several other methods also use this judgment, only as a workaround?
https://github.com/ceph/ceph/blob/f71af8125129ebaa8a02712ca9d6e45fb9a4cd12/src/osd/PG.cc#L490-L498

**Pull request closed without merging**